src/table/dtable.jl

```diff
@@ -103,56 +105,82 @@ instance of table type created using the provided `sink` function.
 """
 fetch(d::DTable, sink) = sink(_retrieve_partitions(d))
 
-function _retrieve_partitions(d::DTable) 
-    trim!(d)
-    return length(d.chunks) > 0 ?
-        TableOperations.joinpartitions(Tables.partitioner(_retrieve, d.chunks)) : NamedTuple()
+function _retrieve_partitions(d::DTable)
+    d2 = trim(d)
+    return length(d2.chunks) > 0 ?
+        TableOperations.joinpartitions(Tables.partitioner(_retrieve, d2.chunks)) : NamedTuple()
 end
 
 _retrieve(x::Dagger.EagerThunk) = fetch(x)
 _retrieve(x::Dagger.Chunk) = collect(x)
 
+"""
+    tabletype!(d::DTable)
+
+Provides the type of the underlying table partition and caches it in `d`.
+
+In case the tabletype cannot be obtained the default return value is `NamedTuple`.
+"""
+tabletype!(d::DTable) = d.tabletype = tabletype(d)
+
 """
     tabletype(d::DTable)
 
 Provides the type of the underlying table partition.
+
+In case the tabletype cannot be obtained the default return value is `NamedTuple`.
 """
 function tabletype(d::DTable)
-    if d.tabletype === nothing
```

Comment:
Why remove this logic? Wouldn't we want to return the internal tabletype if one is already set? It would remove the need for the `_tabletype` logic I think.

Reply:
Ok, I had the assumption in mind that there might be a case where you wouldn't want to use the cached tabletype, because for whatever reason it might be wrong, but now that I think about it there should be no case when it's actually wrong
I'll add the _tabletype logic to tabletype and remove _tabletype then.
I'll also make the tabletype! actually resolve the tabletype every time, because right now it would use the cached one to cache it.